hw/bsp/msp_exp430f5529lp/msp_exp430f5529lp.c

```diff
@@ -0,0 +1,205 @@
+  P5SEL |= BIT2; // Required to enable XT2.
+  // Loop until XT2 fault flag is cleared
+  do
+  {
+    UCSCTL7 &= ~XT2OFFG;
+  }while(UCSCTL7 & XT2OFFG);
+
+  // Kickstart the DCO into the correct frequency range, otherwise a
+  // fault will occur.
+  // FIXME: DCORSEL_6 should work according to datasheet params, but generates
+  // a fault. I am not sure why it faults.
+  UCSCTL1 = DCORSEL_7;
+  UCSCTL2 = FLLD_2 + 3; // DCO freq = D * (N + 1) * (FLLREFCLK / n)
+                        // DCOCLKDIV freq = (N + 1) * (FLLREFCLK / n)
+                        // N = 3, D = 2, thus DCO freq = 32 MHz.
+
+  // MCLK configured for 16 MHz using XT2.
+  // SMCLK configured for 8 MHz using XT2.
+  UCSCTL4 |= SELM__DCOCLKDIV + SELS__DCOCLKDIV;
+  UCSCTL5 |= DIVM__16 + DIVS__2;
+
+  // Now wait till everything's stabilized.
+  do
+  {
+    UCSCTL7 &= ~(XT2OFFG + XT1LFOFFG + DCOFFG);
+    SFRIFG1 &= ~OFIFG;
+  }while(SFRIFG1 & OFIFG);
+
+  // Configure Timer A to use SMCLK as a source. Count 1000 ticks at 1 MHz.
+  TA0CCTL0 |= CCIE;
+  TA0CCR0 = 999; // 1000 ticks.
+  TA0CTL |= TASSEL_2 + ID_3 + MC__UP; // Use SMCLK, divide by 8, start timer.
+
+  // Initialize USB power and PLL.
+  USBKEYPID = USBKEY;
+
+  // VUSB enabled automatically.
+  // Wait two milliseconds to stabilize, per manual recommendation.
+  uint32_t ms_elapsed = board_millis();
+  do
+  {
+    while((board_millis() - ms_elapsed) < 2);
+  }while(!(USBPWRCTL & USBBGVBV));
+
+  // USB uses XT2 (4 MHz) directly. Enable the PLL.
+  USBPLLDIVB |= USBPLL_SETCLK_4_0;
+  USBPLLCTL |= (UPFDEN | UPLLEN);
+
+  // Wait until PLL locks. Check every 2ms, per manual.
+  ms_elapsed = board_millis();
+  do
+  {
+    USBPLLIR &= ~USBOOLIFG;
+    while((board_millis() - ms_elapsed) < 2);
+  }while(USBPLLIR & USBOOLIFG);
+
+  USBKEYPID = 0;
+}
+
+uint32_t wait = 0;
+
+void board_init(void)
+{
+  __bis_SR_register(GIE); // Enable interrupts.
+  SystemClock_Config();
+
+  // Enable basic I/O.
+  P1DIR |= LED_PIN; // LED output.
+  P1REN |= BUTTON_PIN; // Internal resistor enable.
+  P1OUT |= BUTTON_PIN; // Pullup.
+
+  // Enable the backchannel UART (115200)
+  P4DIR |= BIT5;
+  P4SEL |= (BIT5 | BIT4);
+
+  UCA1CTL1 |= (UCSSEL__SMCLK | UCSWRST); // Hold in reset, use SMCLK.
+  UCA1BRW = 4;
+  UCA1MCTL |= (UCBRF_3 | UCBRS_5 | UCOS16); // Overampling mode, 115200 baud.
+                                            // Copied from manual.
+  UCA1CTL1 &= ~UCSWRST;
+
+  // Set up USB pins.
+  USBKEYPID = USBKEY;
+  USBPHYCTL |= PUSEL; // Convert USB D+/D- pins to USB functionality.
+  USBKEYPID = 0;
+}
+
+//--------------------------------------------------------------------+
+// Board porting API
+//--------------------------------------------------------------------+
+
+void board_led_write(bool state)
+{
+  if(state)
+  {
+    LED_PORT |= LED_PIN;
+  }
+  else
+  {
+    LED_PORT &= ~LED_PIN;
+  }
+}
+
+uint32_t board_button_read(void)
+{
+  return (P1IN & BIT1);
```

Comment:
this should be `return ((P1IN & BIT1) >> 1) == BUTTON_STATE_ACTIVE;`
To test this you could run the board_test example, the example does nothing but test the led and button. LED blink normal when button is not active, and blink faster when pressing the button.